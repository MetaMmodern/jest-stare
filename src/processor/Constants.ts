import * as chalk from "chalk";

/**
 * Contains Constants constants
 * @export
 * @class Constants
 */
export class Constants {

    /**
     * Name (same as package.json)
     * @static
     * @memberof Constants
     */
    public static readonly NAME = "jest-stare";

    /**
     * Logo for messaging
     * @static
     * @memberof Constants
     */
    public static readonly LOGO = chalk.default.green("**  ") + chalk.default.green("jest") + chalk.default.yellow("-") + chalk.default.red("stare");

    /**
     * Default jest-stare results directory
     * @static
     * @memberof Constants
     */
    public static readonly DEFAULT_RESULTS_DIR = "./" + Constants.NAME;

    /**
     * Main HTML index file for report
     * @static
     * @memberof Constants
     */
    public static readonly MAIN_HTML = "index.html";

    /**
     * jest-stare render JS file
     * @static
     * @memberof Constants
     */
    public static readonly JEST_STARE_JS = "view.js";

    /**
     * Raw JSON response results
     * @static
     * @memberof Constants
     */
    public static readonly RESULTS_RAW = "jest-results.json";

    /**
     * jest-stare custom css
     * @static
     * @memberof Constants
     */
    public static readonly JEST_STARE_CSS = Constants.NAME + ".css";

    /**
     * Template HTML file
     * @static
     * @memberof Constants
     */
    public static readonly TEMPLATE_HTML = "template.html";

    /**
     * Base dir to contain css files
     * @static
     * @memberof Constants
     */
    public static readonly CSS_DIR = "css/";

    /**
     * Base dir to contain js files
     * @static
     * @memberof Constants
     */
    public static readonly JS_DIR = "js/";

    /**
     * Log content after logo
     * @static
     * @memberof Constants
     */
    public static readonly LOG_MESSAGE = " --testResultsProcessor: wrote output report to ";
}
