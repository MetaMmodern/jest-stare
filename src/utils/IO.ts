import * as fs from "fs";
import * as  mkdirp from "mkdirp";
const pkgUp = require("pkg-up");

/**
 * class for simple IO
 * @export
 * @class IO
 */
export class IO {

    /**
     * Delete a file if it exists
     * @static
     * @param {string} file - file to delete
     * @memberof IO
     */
    public static unlinkSync(file: string) {
        if (IO.existsSync(file)) {
            fs.unlinkSync(file);
        }
    }

    /**
     * Wrap fs.writeFile for promise-based calling
     * @static
     * @param {string} path - place to write to
     * @param {*} data - content to write
     * @returns {Promise<void>} - when write is complete
     * @memberof IO
     */
    public static writeFile(path: string, data: any): Promise<void> {
        return new Promise<void>( (resolve, reject) => {
            fs.writeFile(path, data, (error) => {
                if (error) {
                    reject(error);
                }
                resolve();
            });
        });
    }

    /**
     * Wrap fs.writeFile for promise-based calling
     * @static
     * @param {string} path - place to write to
     * @param {*} data - content to write
     * @returns {Promise<void>} - when write is complete
     * @memberof IO
     */
    public static writeFileSync(path: string, data: any) {
        fs.writeFileSync(path, data);
    }

    /**
     * Make directories (recursive)
     * @static
     * @param {string} dir - directory to make
     * @memberof IO
     */
    public static mkdirsSync(dir: string) {
        if (!IO.existsSync(dir)) {
            mkdirp.sync(dir);
        }
    }

    /**
     * Read file and return to caller
     * @static
     * @param {string} path - path to read file
     * @returns - file contents
     * @memberof IO
     */
    public static readFileSync(path: string) {
        return fs.readFileSync(path).toString();
    }

    /**
     * Return whether a given file already exists
     * @static
     * @param {string} path - path of file to check
     * @returns - true if file exists, false otherwise
     * @memberof IO
     */
    public static existsSync(path: string) {
        return fs.existsSync(path);
    }

    /**
     * Read from the user's package.json, if present
     * @static
     * @returns {IJestStareConfig} - config object
     * @returns {object} - full package JSON
     * @memberof IO
     */
    public static readPackageJson(): object {
        const packageJson = pkgUp.sync();
        if (packageJson !== null) {
            return JSON.parse(IO.readFileSync(packageJson));
        } else {
            // if we can't find any package.json, return a blank config
            return {};
        }
    }
}
