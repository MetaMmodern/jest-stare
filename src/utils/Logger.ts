import { format, isNullOrUndefined } from "util";
import * as moment from "moment";

export class Logger {

    public static readonly LEVELS = ["trace", "debug", "info", "warn", "error", "fatal"];
    public static readonly LEVEL_DEFAULT = "debug";

    public static get get() {
        if (isNullOrUndefined(this.mLog)) {
            this.mLog = new Logger();
        }
        return this.mLog;
    }

    public static isValidLevel(level: string) {
        return Logger.LEVELS.indexOf(level) < 0 ? false : true;
    }

    public static validateLevel(level: string) {
        if (!Logger.isValidLevel(level)) {
            throw new Error("invalid level");
        }
    }

    private static mLog: Logger;
    private mPrefix: boolean;
    private mColor: boolean;
    private mIsOn: boolean;
    private mLevel: string;

    constructor() {
        this.mLevel = Logger.LEVEL_DEFAULT;
        this.mPrefix = true;
        this.mColor = true;
        this.mLevel = this.mLevel.toLocaleLowerCase();
        this.mIsOn = true;
        Logger.validateLevel(this.mLevel);
    }

    public isDebugEnabled() {
        return Logger.LEVELS.indexOf("debug") >= Logger.LEVELS.indexOf(this.level) ? true : false;
    }

    public debug(message: string, ...args: any[]) {
        if (!this.isDebugEnabled()) {
            return;
        }
        let adjustedMessage = message;
        if (this.prefix) {
            adjustedMessage = this.buildPrefix("DEBUG") + message;
        }
        if (this.color) {
            // adjustedMessage = TextUtils.chalk.blue(adjustedMessage);
        }
        return this.writeStdout(adjustedMessage, args);
    }

    private writeStderr(message: string, ...args: any[]) {
        const data = this.format(message, args);
        if (this.on) {
            process.stderr.write(this.format(message, args));
        }
        return data;
    }

    private writeStdout(message: string, ...args: any[]) {
        const data = this.format(message, args);
        if (this.on) {
            process.stdout.write(data);
        }
        return data;
    }

    private format(data: string, ...args: any[]) {
        let formatted = data;
        // TODO(Kelosky): this is not ideal, but works for simple cases of
        // .debug(%s, "sub string").
        if (this.formatEnabled && !isNullOrUndefined(args) && args.length > 0) {
            let defined = false;
            args.forEach((arg) => {
                arg.forEach((ntry: string[]) => {
                    if (ntry.length > 0) {
                        defined = true;
                    }
                });
            });
            // if every argument is undefined, dont format it
            if (defined) {
                formatted = format(data, args);
            }
        }
        return formatted + "\n";
    }

    private buildPrefix(type: string) {
        return "[" + moment().format("YYYY/MM/DD HH:MM:SS") + "]" + " " + "[" + type + "]" + " ";
    }

    get formatEnabled() {
        return true;
    }

    set color(isEnabled: boolean) {
        this.mColor = isEnabled;
    }

    get color(): boolean {
        return this.mColor;
    }

    set on(isOn: boolean) {
        this.mIsOn = isOn;
    }

    get on() {
        return this.mIsOn;
    }

    get level() {
        return this.mLevel;
    }

    get prefix(): boolean {
        return this.mPrefix;
    }

    set prefix(prefix) {
        this.mPrefix = prefix;
    }
}
