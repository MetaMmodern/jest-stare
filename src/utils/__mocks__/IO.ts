import * as fs from "fs";

/**
 * class for simple IO
 * @export
 * @class IO
 */
export class IO {

    /**
     * Wrap fs.writeFile for promise-based calling
     * @static
     * @param {string} path - place to write to
     * @param {*} data - content to write
     * @returns {Promise<void>} - when write is complete
     * @memberof IO
     */
    public static writeFile(path: string, data: any): Promise<void> {
        return new Promise<void>((resolve, reject) => {
            resolve();
        });
    }

    /**
     * Make directories (recursive)
     * @static
     * @param {string} dir - directory to make
     * @memberof IO
     */
    public static mkdirsSync(dir: string) {
        return;
    }

    /**
     * Read file and return to caller
     * @static
     * @param {string} path - path to read file
     * @returns - file contents
     * @memberof IO
     */
    public static readFileSync(path: string) {
        throw new Error("Mock read not implemented yet");
    }

    /**
     * Return whether a given file already exists
     * @static
     * @param {string} path - path of file to check
     * @returns - true if file exists, false otherwise
     * @memberof IO
     */
    public static existsSync(path: string) {
        return true;
    }
}
