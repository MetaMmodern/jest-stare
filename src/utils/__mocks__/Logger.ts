import { isNullOrUndefined } from "util";

const stdout = jest.fn<string>((msg: string) => {
    return msg;
});

const stderr = jest.fn<string>((msg: string) => {
    return msg;
});

export class Logger {
    private static mLog: Logger;

    public static get get() {
        if (isNullOrUndefined(this.mLog)) {
            this.mLog = new Logger();
        }
        return this.mLog;
    }

    private writeStdout = stdout;

    private writeStderr = stderr;

    public debug(message: string, ...args: any[]): string {
        return this.writeStdout(message);
    }

    public error(message: string, ...args: any[]): string {
        return this.writeStderr(message);
    }
}
