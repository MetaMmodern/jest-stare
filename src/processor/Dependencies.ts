import { IThirdPartyDependency } from "./doc/IThirdPartyDependency";
import { Constants } from "./Constants";

/**
 * Describe all third party dependencies - note, these must still be added to template manually for now
 * @export
 * @class Dependencies
 */
export class Dependencies {

    /**
     * All third party dependencies are here
     * @static
     * @type {IThirdPartyDependency[]}
     * @memberof Dependencies
     */
    public static readonly THIRD_PARTY_DEPENDENCIES: IThirdPartyDependency[] = [

        // js
        {
            requireDir: "bootstrap/dist/js/",
            file: "bootstrap.min.js",
            targetDir: Constants.JS_DIR
        },
        {
            requireDir: "diff2html/dist/",
            file: "diff2html.min.js",
            targetDir: Constants.JS_DIR
        },
        {
            requireDir: "jquery/dist/",
            file: "jquery.min.js",
            targetDir: Constants.JS_DIR
        },
        {
            requireDir: "holderjs/",
            file: "holder.js",
            targetDir: Constants.JS_DIR
        },

        // css
        {
            requireDir: "bootstrap/dist/css/",
            file: "bootstrap.min.css",
            targetDir: Constants.CSS_DIR
        },
        {
            requireDir: "diff2html/dist/",
            file: "diff2html.min.css",
            targetDir: Constants.CSS_DIR
        },
    ];
}
