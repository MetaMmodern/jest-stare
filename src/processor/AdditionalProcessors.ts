import { IResultsProcessorInput } from "./doc/jest/IResultsProcessorInput";
import { Logger } from "../utils/Logger";
import { Constants } from "./Constants";


export class AdditionalProcessors {

    /**
     * Pass the result processor input given to jest-stare to additional
     * test results processors
     * @param jestTestData - input passed to jest-stare
     * @param processors - list of test results processors (e.g. ["jest-html-reporter"])
     *                     to forward the data to
     */
    public static execute(jestTestData: IResultsProcessorInput, processors: string[]): void {
        const chalk = require("chalk");
        for (const processor of processors) {
            try {
                require(processor)(jestTestData);
                AdditionalProcessors.logger.debug(Constants.LOGO + " passed results to additional processor \"" +
                    processor + "\"" + chalk.default.green("\t**"));
            } catch (e) {
                AdditionalProcessors.logger.error("Error executing additional processor: \"" + processor + "\"" + e);
            }
        }
    }

    private static logger = new Logger();
}
