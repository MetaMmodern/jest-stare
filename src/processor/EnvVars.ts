import { IJestStareConfig } from "./doc/IJestStareConfig";
import { Logger } from "../utils/Logger";
import { EnvVarService } from "../utils/EnvVarService";

/**
 * Service for reading environmental variables
 * @export
 * @class EnvVars
 */
export class EnvVars {

    /**
     * Prefix for all jest-stare related environmental variables
     * @static
     * @type {string}
     * @memberof EnvVars
     */
    public static readonly ENV_PREFIX: string = "JEST_STARE_";

    /**
     * Creates an instance of EnvVars.
     * @memberof EnvVars
     */
    constructor(private mEnvSrv = new EnvVarService(EnvVars.ENV_PREFIX)) {}

    /**
     * Read all jest-stare related environmental variables
     * @returns {IJestStareConfig} - jest stare config from env vars
     * @memberof EnvVars
     */
    public read(): IJestStareConfig {
        const additionalResultsProcessorsValue = this.mEnvSrv.readEnvValue("ADDITIONAL_RESULTS_PROCESSORS");
        let additionalResultsProcessors: string[];
        if (additionalResultsProcessorsValue != null) {
            try {
                additionalResultsProcessors = JSON.parse(additionalResultsProcessorsValue);
            } catch (e) {
                Logger.get.error("Could not parse additional results processors value." +
                    "It should be a JSON string of an array of strings, like the following: " +
                    JSON.stringify(["jest-html-reporter"]) + "\n You specified: " + additionalResultsProcessorsValue);
            }
        }
        return {
            resultDir: this.mEnvSrv.readEnvValue("RESULT_DIR"),
            resultJson: this.mEnvSrv.readEnvValue("RESULT_JSON"),
            resultHtml: this.mEnvSrv.readEnvValue("RESULT_HTML"),
            log: this.mEnvSrv.readBoolEnvValue("LOG"),
            merge: this.mEnvSrv.readBoolEnvValue("MERGE"),
            jestStareConfigJson: this.mEnvSrv.readEnvValue("CONFIG_JSON"),
            coverageLink: this.mEnvSrv.readEnvValue("COVERAGE_LINK"),
            additionalResultsProcessors
        };
    }

    /**
     * Combine jest stare config from package json and environmental variables,
     * with environmental variables taking precedence
     * @param packageJsonConfig
     * @param {IJestStareConfig} packageJsonConfig - package json config
     * @param {IJestStareConfig} envConfig - env based config
     * @returns {IJestStareConfig} - merged config
     * @memberof EnvVars
     */
    public resolve(packageJsonConfig: IJestStareConfig, envConfig: IJestStareConfig): IJestStareConfig {
        const mergedConfig: IJestStareConfig = {};

        if (envConfig.resultDir != null || packageJsonConfig.resultDir != null) {
            mergedConfig.resultDir = envConfig.resultDir == null ? packageJsonConfig.resultDir : envConfig.resultDir;
        }

        if (envConfig.resultJson != null || packageJsonConfig.resultJson != null) {
            mergedConfig.resultJson = envConfig.resultJson == null ? packageJsonConfig.resultJson : envConfig.resultJson;
        }

        if (envConfig.resultHtml != null || packageJsonConfig.resultHtml != null) {
            mergedConfig.resultHtml = envConfig.resultHtml == null ? packageJsonConfig.resultHtml : envConfig.resultHtml;
        }

        if (envConfig.log != null || packageJsonConfig.log != null) {
            mergedConfig.log = envConfig.log == null ? packageJsonConfig.log : envConfig.log;
        }

        if (envConfig.merge != null || packageJsonConfig.merge != null) {
            mergedConfig.merge = envConfig.merge == null ? packageJsonConfig.merge : envConfig.merge;
        }

        if (envConfig.jestStareConfigJson != null || packageJsonConfig.jestStareConfigJson != null) {
            mergedConfig.jestStareConfigJson =
                envConfig.jestStareConfigJson == null ? packageJsonConfig.jestStareConfigJson : envConfig.jestStareConfigJson;
        }


        if (envConfig.coverageLink != null || packageJsonConfig.coverageLink != null) {
            mergedConfig.coverageLink = envConfig.coverageLink == null ? packageJsonConfig.coverageLink : envConfig.coverageLink;
        }

        if (envConfig.additionalResultsProcessors != null || packageJsonConfig.additionalResultsProcessors != null) {
            mergedConfig.additionalResultsProcessors =
                envConfig.additionalResultsProcessors == null ? packageJsonConfig.additionalResultsProcessors : envConfig.additionalResultsProcessors;
        }

        return mergedConfig;
    }
}
