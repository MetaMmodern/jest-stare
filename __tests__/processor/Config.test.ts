import { Config } from "../../src/processor/Config";
import { Logger } from "../../src/utils/Logger";
import { IO } from "../../src/utils/IO";

describe("Config tests", () => {
    it("should be able to read a config", () => {
        IO.readPackageJson = jest.fn(() => {
            return {
                "jest-stare": {
                    additionalResultsProcessors: [
                        "jest-html-reporter",
                        "jestjunit",
                        "also-one-that-doesnt-exist"
                    ],
                    coverageLink: "../../coverage/lcov-report/index.html",
                    jestStareConfigJson: "jest-stare.json",
                    merge: "true",
                    resultDir: "results/jest-stare/",
                    resultHtml: "index.html",
                    resultJson: "jest-results.json"
                }
            };
        });
        const response = new Config(new Logger(), undefined, undefined).buildConfig();
        expect(response).toMatchSnapshot();
    });
});