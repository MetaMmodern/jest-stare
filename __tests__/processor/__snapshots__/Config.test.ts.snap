// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config tests should be able to read a config 1`] = `
Object {
  "additionalResultsProcessors": Array [
    "jest-html-reporter",
    "jestjunit",
    "also-one-that-doesnt-exist",
  ],
  "coverageLink": "../../coverage/lcov-report/index.html",
  "jestStareConfigJson": "jest-stare.json",
  "resultDir": "results/jest-stare//",
  "resultHtml": "index.html",
  "resultJson": "jest-results.json",
}
`;
